crates/solvers/src/boundary/naive.rs

```diff
@@ -0,0 +1,239 @@
+        .iter()
+        .map(|order| LimitOrder {
+            id: match order.class {
+                order::Class::Market => LimitOrderId::Market(OrderUid(order.uid.0)),
+                order::Class::Limit => LimitOrderId::Limit(OrderUid(order.uid.0)),
+                order::Class::Liquidity => {
+                    LimitOrderId::Liquidity(LiquidityOrderId::Protocol(OrderUid(order.uid.0)))
+                }
+            },
+            sell_token: order.sell.token.0,
+            buy_token: order.buy.token.0,
+            sell_amount: order.sell.amount,
+            buy_amount: order.buy.amount,
+            kind: match order.side {
+                order::Side::Buy => OrderKind::Buy,
+                order::Side::Sell => OrderKind::Sell,
+            },
+            partially_fillable: order.partially_fillable,
+            solver_fee: order.fee().amount,
+            settlement_handling: Arc::new(OrderHandler {
+                order: Order {
+                    metadata: OrderMetadata {
+                        uid: OrderUid(order.uid.0),
+                        class: match order.class {
+                            order::Class::Market => OrderClass::Market,
+                            order::Class::Limit => OrderClass::Limit(LimitOrderClass {
+                                surplus_fee: Some(order.fee().amount),
+                                ..Default::default()
+                            }),
+                            order::Class::Liquidity => OrderClass::Liquidity,
+                        },
+                        solver_fee: order.fee().amount,
+                        ..Default::default()
+                    },
+                    data: OrderData {
+                        sell_token: order.sell.token.0,
+                        buy_token: order.buy.token.0,
+                        sell_amount: order.sell.amount,
+                        buy_amount: order.buy.amount,
+                        fee_amount: order.fee().amount,
+                        kind: match order.side {
+                            order::Side::Buy => OrderKind::Buy,
+                            order::Side::Sell => OrderKind::Sell,
+                        },
+                        partially_fillable: order.partially_fillable,
+                        ..Default::default()
+                    },
+                    ..Default::default()
+                },
+            }),
+            exchange: Exchange::GnosisProtocol,
+            reward: 0.,
+        })
+        .collect_vec();
+
+    let slippage = Slippage::new(pool.tokens());
+    let pool_handler = Arc::new(PoolHandler::default());
+    let boundary_pool = ConstantProductOrder::for_pool(
+        to_boundary_pool(liquidity.address, pool)?,
+        pool_handler.clone(),
+    );
+
+    let boundary_solution =
+        multi_order_solver::solve(&slippage.context(), boundary_orders, &boundary_pool)?;
+
+    let mut swap = pool_handler.swap.lock().unwrap();
```

Comment:
nit: Not that it matters here (since this Mutex only gets accessed from the same thread) but I think it would be nice to avoid holding locks if possible.

```suggestion
    let swap = pool_handler.swap.lock().unwrap().take();
```

Reply:
Thanks. I think these are good defensive programming suggestions - will definitely apply them!

Reply:
6700e0fb429776121c5399e4388c24e763c29367